src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java

```diff
@@ -177,4 +186,119 @@ public ExamResource(UserService userService, ExamService examService, ExamAccess
+     * @param students     the list of students (with at least registration number) who should get access to the exam
+     * @return             the list of students who could not be registered for the exam, because they could NOT be found in the Artemis database and could NOT be found in the TUM LDAP
+     */
+    @PostMapping(value = "/courses/{courseId}/exams/{examId}/students")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<List<StudentDTO>> addStudentToCourse(@PathVariable Long courseId, @PathVariable Long examId, @RequestBody List<StudentDTO> students) {
+        log.debug("REST request to add {} as students to exam : {}", students, examId);
+        var course = courseService.findOne(courseId);
+        var instructorOrAdmin = userService.getUserWithGroupsAndAuthorities();
+        if (!authCheckService.isAtLeastInstructorInCourse(course, instructorOrAdmin)) {
+            return forbidden();
+        }
+        var exam = examService.findOneWithRegisteredUsers(examId);
+        if (!course.equals(exam.getCourse())) {
+            return forbidden();
+        }
+        List<StudentDTO> notFoundStudents = new ArrayList<>();
+        for (var student : students) {
+            var registrationNumber = student.getRegistrationNumber();
+            // 1) we use the registration number and try to find the student in the Artemis user database
+            Optional<User> user = userService.findUserWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumber);
+            if (user.isPresent()) {
+                exam.addUser(user.get());
+                // we only need to add the user to the course group, if the student is not yet part of it, otherwise the student cannot access the exam (within the course)
+                if (!user.get().getGroups().contains(course.getStudentGroupName())) {
+                    userService.addUserToGroup(user.get(), course.getStudentGroupName());
+                }
+                continue;
+            }
+            // 2) if we cannot find the user, we use the registration number and try to find the student in the (TUM) LDAP
+            user = userService.createUserFromLdap(registrationNumber);
+            if (user.isPresent()) {
+                exam.addUser(user.get());
+                // the newly created user needs to get the rights to access the course, otherwise the student cannot access the exam (within the course)
+                userService.addUserToGroup(user.get(), course.getStudentGroupName());
+                continue;
+            }
+            // 3) if we cannot find the user in the (TUM) LDAP, we report this to the client
+            log.warn("User with registration number " + registrationNumber + " not found in Artemis user database and not found in (TUM) LDAP");
+            notFoundStudents.add(student);
+        }
+        examRepository.save(exam);
+        return ResponseEntity.ok().body(notFoundStudents);
+    }
```

Comment:
Could this be an improvement? I think you can cut the for loop there and improve the performance through intelligend Database queries instead of tackeling this one by one.
Don't know for sure though. Would be interesting what you think.

Keep in mind this was only written in Github not with a proper IDE and I invented some "multiple count"-variations of already existing methods were I thought it would help the performance.
```suggestion
        List<StudentDTO> notFoundStudents = new ArrayList<>();
        List<String> registrationNumbers = students.stream().map(StudentDTO::getRegistrationNumber()).collect(Collections.toList());
        List<User> artemisUsers = userService.findAllUsersWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumbers);
        exam.addAllUsers(artemisUsers);
        
        //Get only registration numbers in here that are not present in artemis
        registrationNumbers.removeAll(artemisUsers.stream().map(User::getRegistrationNumber()).collect(Collections.toList());
        List<User> artemisUsersFromLdap = registrationNumbers.stream().map(e->userService.createUserFromLdap(e)).collect(Collections.toList());
        exam.addAllUsers(artemisUsersFromLdap);
        userService.addUsersToGroup(artemisUsersFromLdap, course.getStudentGroupName());
        
        //Get only registration numbers in here that were not found at all
        registrationNumbers.removeAll(artemisUsersFromLdap.stream().map(User::getRegistrationNumber()).collect(Collections.toList());
        
        //ToDo: print proper warning
            // 3) if we cannot find the user in the (TUM) LDAP, we report this to the client
            log.warn("User with registration number " + registrationNumber + " not found in Artemis user database and not found in (TUM) LDAP");
        examRepository.save(exam);
        
        return ResponseEntity.ok().body(students.stream.filter(e->registrationNumbers.contains(e.getRegistrationNumber())).collect(Collections.toList()));
    }
```

Reply:
Thanks for the detailed comment. This functionality won't be used often, so I keep the slower (but in my opinion easier to understand) version for now. I also added some additional source code and reworking it would be too much effort for now.